src/fs.rs

```diff
@@ -0,0 +1,419 @@
+use crate::{
+    common::{str2wcstring, wcs2zstring, ScopeGuard},
+    fds::wopen_cloexec,
+    global_safety::RelaxedAtomicBool,
+    path::{path_get_data_remoteness, DirRemoteness},
+    wchar::prelude::*,
+    wutil::{file_id_for_file, file_id_for_path, wrename, wunlink, FileId, INVALID_FILE_ID},
+    FLOG,
+};
+use errno::errno;
+use libc::{c_int, fchown, flock, EINTR, LOCK_EX, LOCK_SH, LOCK_UN};
+use nix::{fcntl::OFlag, sys::stat::Mode};
+use std::{
+    ffi::CString,
+    fs::File,
+    io::Seek,
+    os::{
+        fd::{AsRawFd, FromRawFd},
+        unix::fs::MetadataExt,
+    },
+};
+
+// Replacement for mkostemp(str, O_CLOEXEC)
+// This uses mkostemp if available,
+// otherwise it uses mkstemp followed by fcntl
+fn fish_mkstemp_cloexec(name_template: CString) -> std::io::Result<(File, CString)> {
+    let name = name_template.into_raw();
+    #[cfg(not(apple))]
+    let fd = {
+        use libc::O_CLOEXEC;
+        unsafe { libc::mkostemp(name, O_CLOEXEC) }
+    };
+    #[cfg(apple)]
+    let fd = {
+        use libc::{FD_CLOEXEC, F_SETFD};
+        let fd = unsafe { libc::mkstemp(name) };
+        if fd != -1 {
+            unsafe { libc::fcntl(fd, F_SETFD, FD_CLOEXEC) };
+        }
+        fd
+    };
+    if fd == -1 {
+        Err(std::io::Error::from(errno()))
+    } else {
+        unsafe { Ok((File::from_raw_fd(fd), CString::from_raw(name))) }
+    }
+}
+
+/// Creates a temporary file created according to the template and its name if successful.
+pub fn create_temporary_file(name_template: &wstr) -> std::io::Result<(File, WString)> {
+    let (fd, c_string_template) = loop {
+        match fish_mkstemp_cloexec(wcs2zstring(name_template)) {
+            Ok(tmp_file_data) => break tmp_file_data,
+            Err(e) => match e.kind() {
+                std::io::ErrorKind::Interrupted => {}
```

Comment:
yeah this looks better.
I agree that the existing retries are probably for EINTR and the retry-limit
seems like an awkward workaround to avoid an infinite loop on non-retriable errors..